src/main/scala/cromwell/engine/workflow/WorkflowMetadataBuilder.scala

```diff
@@ -0,0 +1,164 @@
+    }
+    // TODO: Refactor db to expose the clob <-> string implicits.
+    val workflowInputs = Source.fromInputStream(workflowExecutionAux.jsonInputs.getAsciiStream)
+      .mkString.parseJson.asInstanceOf[JsObject]
+
+    WorkflowMetadataResponse(
+      id = workflowExecution.workflowExecutionUuid.toString,
+      workflowName = workflowExecution.name,
+      status = workflowExecution.status,
+      // We currently do not make a distinction between the submission and start dates of a workflow, but it's
+      // possible at least theoretically that a workflow might not begin to execute immediately upon submission.
+      submission = startDate,
+      start = Option(startDate),
+      end = endDate,
+      inputs = workflowInputs,
+      outputs = Option(workflowOutputs) map {
+        _.mapToValues
+      },
+      calls = callMetadata)
+  }
+
+  private def buildWorkflowMetadata(id: WorkflowId,
+                                    backend: Backend,
+                                    workflowExecution: WorkflowExecution,
+                                    workflowOutputs: Traversable[SymbolStoreEntry],
+                                    workflowExecutionAux: WorkflowExecutionAux,
+                                    workflowDescriptor: WorkflowDescriptor,
+                                    executionStatuses: Map[db.ExecutionDatabaseKey, db.CallStatus],
+                                    callInputs: Traversable[SymbolStoreEntry],
+                                    callOutputs: Traversable[SymbolStoreEntry],
+                                    infosByExecution: Traversable[ExecutionInfosByExecution],
+                                    runtimeAttributes: Map[ExecutionDatabaseKey, Map[String, String]],
+                                    executionEvents: Map[db.ExecutionDatabaseKey, Seq[engine.ExecutionEventEntry]]):
+  WorkflowMetadataResponse = {
+    val nonFinalEvents = executionEvents.filterKeys(!_.fqn.isFinalCall)
+    val nonFinalStatuses = executionStatuses.filterKeys(!_.fqn.isFinalCall)
+    val nonFinalInfosByExecution = infosByExecution.filterNot(_.execution.callFqn.isFinalCall)
+
+    val engineWorkflowOutputs = SymbolStoreEntry.toWorkflowOutputs(workflowOutputs)
+    val callStandardStreamsMap = workflowStdoutStderr(id, backend, workflowDescriptor, nonFinalStatuses)
+    val callMetadata = CallMetadataBuilder.build(nonFinalInfosByExecution, callStandardStreamsMap, callInputs,
+      callOutputs, nonFinalEvents, runtimeAttributes)
+    buildWorkflowMetadata(workflowExecution, workflowExecutionAux, engineWorkflowOutputs, callMetadata)
+  }
+
+  def workflowMetadata(id: WorkflowId, backend: Backend)
+                      (implicit ec: ExecutionContext): Future[WorkflowMetadataResponse] = {
+
+    // TODO: This entire block of chained database actions should be a single request to the db layer.
+    // Also, because it's read only, this would be a candidate for a separate, optimized data service.
```

Comment:
Removing the second comment, but it was a just a note that we have room for optimization our database interactions. Specifically, we could have a separate readonly database accessor with its own thread pool.